src/Microsoft.Extensions.Configuration.AzureAppConfiguration/TracingUtils.cs

```diff
@@ -95,16 +106,57 @@ public static async Task CallWithRequestTracing(bool tracingEnabled, RequestType
             }
         }
 
-        private static void AddRequestType(IList<KeyValuePair<string, string>> correlationContext, RequestType requestType)
+        private static string CreateCorrelationContextHeader(RequestType requestType, RequestTracingOptions requestTracingOptions)
         {
-            string requestTypeValue = Enum.GetName(typeof(RequestType), requestType);
-            correlationContext.Add(new KeyValuePair<string, string>(RequestTracingConstants.RequestTypeKey, requestTypeValue));
-        }
+            IList<KeyValuePair<string, string>> correlationContextKeyValues = new List<KeyValuePair<string, string>>();
+            IList<string> correlationContextTags = new List<string>();
+            
+            correlationContextKeyValues.Add(new KeyValuePair<string, string>(RequestTracingConstants.RequestTypeKey, Enum.GetName(typeof(RequestType), requestType)));
 
-        private static void AddHostType(IList<KeyValuePair<string, string>> correlationContext, HostType hostType)
-        {
-            string hostTypeValue = Enum.GetName(typeof(HostType), hostType);
-            correlationContext.Add(new KeyValuePair<string, string>(RequestTracingConstants.HostTypeKey, hostTypeValue));
+            if (requestTracingOptions.HostType != HostType.Unidentified)
+            {
+                correlationContextKeyValues.Add(new KeyValuePair<string, string>(RequestTracingConstants.HostTypeKey, Enum.GetName(typeof(HostType), requestTracingOptions.HostType)));
+            }
+
+            if (requestTracingOptions.IsDevEnvironment)
+            {
+                correlationContextTags.Add(RequestTracingConstants.DevEnvironmentTag);
+            }
+
+            if (requestTracingOptions.IsKeyVaultConfigured)
+            {
+                correlationContextTags.Add(RequestTracingConstants.KeyVaultConfiguredTag);
+            }
+
+            if (requestTracingOptions.IsOfflineCacheConfigured)
+            {
+                correlationContextTags.Add(RequestTracingConstants.OfflineCacheConfiguredTag);
+            }
+
+            string headerKeyValues = "";
+            string headerTags = "";
+            string fullHeader = "";
+
+            if (correlationContextKeyValues.Count > 0)
+            {
+                headerKeyValues = string.Join(",", correlationContextKeyValues.Select(kvp => $"{kvp.Key}={kvp.Value}"));
+            }
+                
+            if (correlationContextTags.Count > 0)
+            {
+                headerTags = string.Join(",", correlationContextTags);
+            }
+
+            if (!string.IsNullOrWhiteSpace(headerKeyValues) && !string.IsNullOrWhiteSpace(headerTags))
+            {
```

Comment:
The if/else took me a second. We have this pattern elsewhere and an alternative way to do it is like the following.

```
StringBuilder sb

foreach (string s in list)
{
    if (sb.Length > 0)
    {
        sb.Append(",");
    }

    sb.Append(s);
}
```

As far as pro's cons, this approach is more of a list building rather than a branched logic.